File: deps/src/polymake_direct_calls.cpp
#include "polymake_includes.h"

#include "polymake_type_modules.h"

template<typename Scalar>
pm::Vector<Scalar> direct_call_solve_LP(
    const pm::Matrix<Scalar>& inequalities,
    const pm::Matrix<Scalar>& equalities,
    const pm::Vector<Scalar>& objective,
    bool                      maximize)
{
    try {
        auto solution = polymake::polytope::solve_LP(inequalities, equalities, objective, maximize);
        return solution.solution;
    } catch (...) {
        return pm::Vector<Scalar>();
    }
}


void polymake_module_add_direct_calls(jlcxx::Module& polymake)
{
    polymake.method("direct_call_solve_LP", &direct_call_solve_LP<pm::Rational>);
    polymake.method("direct_call_solve_LP_float", &direct_call_solve_LP<double>);
}

File: src/polymake_direct_calls.jl
const AbstractRational = Union{Base.Rational, Rational}

for (pm_solve_LP, scalarT, concreteT) in [
    (:direct_call_solve_LP, AbstractRational, Rational),
    (:direct_call_solve_LP_float, AbstractFloat, Float64),
    ]
    for (m, flag) in [(:min, false), (:max, true)]
        @eval begin
            function solve_LP(
                inequalities::AbstractMatrix{<:$scalarT},
                equalities::AbstractMatrix{<:$scalarT},
                objective::AbstractVector{<:$scalarT},
                sense::typeof($m))

                return $pm_solve_LP(
                    convert(Matrix{$concreteT}, inequalities),
                    convert(Matrix{$concreteT}, equalities),
                    convert(Vector{$concreteT}, objective),
                    $flag
                )
            end
        end
    end
end

solve_LP(inequalities::AbstractMatrix{T}, objective::AbstractVector; sense=max) where T = solve_LP(inequalities, Base.Matrix{T}(undef, 0, 0), objective, sense)

solve_LP(inequalities::AbstractMatrix, equalities::AbstractMatrix, objective::AbstractVector; sense=max) = solve_LP(inequalities, equalities, objective, sense)
